src/client/src/producer.rs

```diff
@@ -62,21 +100,76 @@ impl TopicProducer {
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        // Split keys into a vec to feed to the partitioner
+        let (keys, values): (Vec<Option<Vec<u8>>>, Vec<Vec<u8>>) = records
+            .into_iter()
+            .map(|(k, v)| (k.map(|it| it.into()), v.into()))
+            .unzip();
+        let key_slices: Vec<_> = keys.iter().map(|it| it.as_deref()).collect();
+
+        let partitions = {
+            // Ensure we drop partitioner (and its mutex guard) after we're done partitioning
+            let mut partitioner = self.partitioner.lock().await;
+            partitioner.update_config(partition_config);
+            partitioner.partition(&key_slices)
+        };
+
+        // Zip up the partition assignments with their keys and values
+        let iter: Vec<_> = partitions
             .into_iter()
-            .map(|(key, value): (Option<K>, V)| {
-                let key = key.map(|k| Bytes::from(k.into()));
-                let value = Bytes::from(value.into());
-                (key, value)
-            })
+            .zip(keys.into_iter().zip(values.into_iter()))
             .collect();
-        send_records_raw(spu_client, &replica, records).await?;
+
+        // Convert key/values into records and sort them by partition
+        let mut records_by_partition: HashMap<i32, Vec<_>> = HashMap::new();
+        for (partition, (key, value)) in iter {
+            let key = key.map(|it| DefaultAsyncBuffer::new(it));
+            let record = DefaultAsyncBuffer::new(value);
+            let record = DefaultRecord::from((key, record));
+
+            records_by_partition
+                .entry(partition)
+                .or_insert_with(Vec::new)
+                .push(record);
+        }
+
+        // Create one request per partition
+        // TODO collapse this into one request per leader (e.g. multiple partitions per leader)
+        let mut requests = vec![];
+        for (partition, records) in records_by_partition.into_iter() {
+            let mut request = DefaultProduceRequest::default();
+            let mut topic_request = DefaultTopicRequest::default();
+            let mut partition_request = DefaultPartitionRequest::default();
+
+            partition_request.partition_index = partition;
+            partition_request
+                .records
+                .batches
+                .push(DefaultBatch::new(records));
+            topic_request.name = self.topic.clone();
+            topic_request.partitions.push(partition_request);
+            request.acks = 1;
+            request.timeout_ms = 1500;
+            request.topics.push(topic_request);
+
+            requests.push((partition, request));
+        }
+
+        for (partition, request) in requests {
+            let replica = ReplicaKey::new(&self.topic, partition);
+            let mut spu_client = self.pool.create_serial_socket(&replica).await?;
```

Comment:
This seems inefficient, we are creating new client event thought they are same.  Should cache them per SPU given partition

Reply:
Yes, I was thinking about that as well. In order to make that happen, I'll essentially need a map of `SpuId -> [PartitionId]`, so that I could create one request per SPU rather than per partition. I will look today to see how to put that together.

Reply:
Actually, I want a mapping of `PartitionId -> LeaderId`.

I am looking through the `TopicSpec` metadata to try to figure out how to get this mapping, and I have a few questions. 

- I see we can either have an `Assigned` or `Computed` topic, and it seems like the `PartitionMaps` in the  Assigned topic might give me the information that I need
  - I see the `PartitionMaps` is essentially a `[PartitionMap { id: PartitionId, replicas: [ SpuId ] } ]`
  - Does the index into `PartitionMaps` represent the leader ID? It doesn't seem that this would make sense, because then there could only be one partition assigned to a leader.
  - Does the index into `PartitionMaps` actually represent the Partition ID, and the thing called `id: PartitionId` actually represent the leader's SPU ID? I think that would make more sense because then multiple partitions could be assigned to the same leader
- How do we handle Computed topics?

@sehz let me know if I'm on the right track or if I'm way off

Reply:
Topics PartitionMap represents allocations. It tells which replica (unit of storage) is assigned to SPU. It doesn't tell you about leader.  That is tracked in the `Partition` spec.

Reply:
so topic list tells you where replicas are:
```
flvd topic list
 NAME   TYPE      PARTITIONS  REPLICAS  IGNORE-RACK  STATUS                   REASON 
 test   computed      1          1                   resolution::provisioned   
 test2  computed      1          2                   resolution::provisioned   
```

but partition tells you which are leader:
```
flvd partition list
 TOPIC  PARTITION  LEADER  REPLICAS  RESOLUTION  HW    LEO   LSR  FOLLOWER OFFSETS 
 test2  0          5001    [5002]    Online      1     1     0    [ReplicaStatus { spu: 5002, hw: 0, leo: 0 }] 
 test   0          5001    []        Online      1000  1000  0    [] 
```